ussologin/ussologin.go

```diff
@@ -83,41 +92,47 @@ var loginForm = form.Form{
 	},
 }
 
-// LoginWithToken completes a login attempt using tok. The ussoAuthURL
-// should have been obtained from the UbuntuSSOOAuth field in a response
-// to a LoginMethods request from the target service.
-func LoginWithToken(client *http.Client, ussoAuthUrl string, tok *usso.SSOData) error {
-	req, err := http.NewRequest("GET", ussoAuthUrl, nil)
-	if err != nil {
-		return errgo.Notef(err, "cannot create request")
-	}
-	base := *req.URL
-	base.RawQuery = ""
-	rp := usso.RequestParameters{
-		HTTPMethod:      req.Method,
-		BaseURL:         base.String(),
-		Params:          req.URL.Query(),
-		SignatureMethod: usso.HMACSHA1{},
-	}
-	if err := tok.SignRequest(&rp, req); err != nil {
-		return errgo.Notef(err, "cannot sign request")
-	}
-	resp, err := client.Do(req)
-	if err != nil {
-		return errgo.Notef(err, "cannot do request")
+// A TokenGetter is used to fetch a Ubuntu SSO OAuth token.
+type TokenGetter interface {
+	GetToken(context.Context) (*usso.SSOData, error)
+}
+
+// A StoreTokenGetter is a TokenGetter that will try to retrieve the
+// token from some storage, before falling back to another TokenGetter.
+// If the fallback TokenGetter sucessfully retrieves a token then that
+// token will be put in the store.
+type StoreTokenGetter struct {
+	Store       TokenStore
+	TokenGetter TokenGetter
+}
+
+// GetToken implements TokenGetter.GetToken. If a store is specified then
+// GetToken will try to retrieve a token from there. If a stored token is
+// not available then GetToken will fallback to TokenGetter.GetToken (if
+// configured).
+func (g StoreTokenGetter) GetToken(ctx context.Context) (*usso.SSOData, error) {
+	err := errgo.New("store not configured")
+	if g.Store != nil {
```

Comment:
I still don't quite understand why you'd want to use a StoreTokenGetter without a Store. Wouldn't you just use the underlying TokenGetter instead?